### DISCUSSION TITLE
Use tsserver insted of vtsls in lazyVim v12.

### DISCUSSION THREAD

**Discussion by patelharsh9797** [LazyVim/LazyVim#3603] [Q&A]
I want to force lazyvim to use my custom-configured tsserver instead of vtsls. After the recent update lazyvim is behaving so weirdly for me. Especially for my typescript project. It's showing me unnecessary long notes making the whole screen almost unresponsive. Auto complete is f*** up. All the snippets is not working at all. like for example useStateHookSnipet. It shows the ${0} ${1} it's not updating the setState also the same for the rafce snippet for react/next. Not auto jumping. The same thing happened for imd snippet not jumping back and all so I assume that because it's using vtsls?

It might be because of some other reason but I still want to use tsserver instead of vtsls. I tried to remove it from Mason but it's auto-install on new startup.

**Comment by dpetka2001** [comment]
You can't use the current Typescript Extra without `vtsls`. If you want to use `tsserver`, then don't enable the Typescript Extra and use the configuration from [this point in time](https://github.com/LazyVim/LazyVim/blob/54d7cea30567577bf926765e9282d7736e951941/lua/lazyvim/plugins/extras/lang/typescript.lua) which is the last known configuration using `tsserver` before transitioning to `vtsls`. Just put that in some file under `/lua/plugins/` directory. You can skip the lines that have `recommended` as they are not needed in the user personal configuration. 

Regarding the snippets, there are some limitations with native `vim.snippet` engine, so you might want to enable the Luasnip Extra and see if that solves your problem.

**Reply by vdegenne** [reply]
@dpetka2001 I am fine with `vtsls`, the only problem is I don't get completion for function calls. I don't know if that's just a regression or if I need to enable a flag in the configuration. Would you happen to know how?

**Reply by vdegenne** [reply]
oh wait never mind I just had to uncomment the typescript extra >.<

**Comment by patelharsh9797** [comment]
[completion.lua.txt](https://github.com/user-attachments/files/15802206/completion.lua.txt)
[lsp.lua.txt](https://github.com/user-attachments/files/15802207/lsp.lua.txt)

I have this 2 lua file for lsp (tsserver) and the completion (snippet).

How should I enable the "Luasnip Extra".

**Reply by dpetka2001** [reply]
Just open the LazyExtras UI either from Dashboard by pressing `x` or from a normal buffer by pressing `:LazyExtras`. From there just search for the Luasnip Extra and press `x` to enable it. Restart Neovim for changes to take effect.

PS: links to your dotfiles on Github would be more appreciated than having to download files.

**Reply by patelharsh9797** [reply]
Here is the update nvim dotfiles: [nvim config files](https://github.com/patelharsh9797/nvim)

I tried ":LazyExtras" and searched for the Luasnip Extra but there is no "Luasnip Extra" :(

**Reply by dpetka2001** [reply]
Are you sure LazyVim is up to date? This is how it should look like with latest LazyVim version
![2024-06-12_12-38](https://github.com/LazyVim/LazyVim/assets/12776461/ef39e244-dac3-42b4-91a7-9381167d4063)

**Reply by patelharsh9797** [reply]
I guess it is up to date. As snippet are working but not working as it should be like it was working in v10.

like in this example when using `imd` snippet it's say `import { ${2:second} } from "${1:first}"` .
After typing the first and then if I press tab it stays in same scope and tab space in same place instead of jumping to the second argument.
Same thing happen for the `useStateSnippet`.


![image](https://github.com/LazyVim/LazyVim/assets/105432105/8fd432cc-8e83-480b-8d34-9469a7632b1d)


![Screenshot from 2024-06-12 15-12-31](https://github.com/LazyVim/LazyVim/assets/105432105/77435dd7-fa06-4337-b86f-737abbf1c26a)

**Reply by dpetka2001** [reply]
You have not enabled the Luasnip Extra. Just go with your cursor over `coding.luasnip` and press `x` to enable it and restart Neovim.

**Reply by patelharsh9797** [reply]
Thnx mate it's works.
i added this in my lazy.lua so now it would be enable by default.
`{ import = "lazyvim.plugins.extras.coding.luasnip" }`

Any other LazyExtras plugin that should be enabled? just want a recommendation or something.

**Reply by dpetka2001** [reply]
I only have these enabled
![2024-06-12_13-24](https://github.com/LazyVim/LazyVim/assets/12776461/61b0459b-fadc-4eb8-a8ed-522d317d2b35)
It really depends on the user's preferences I believe.

**Reply by patelharsh9797** [reply]
Thnx a lot for helping out. :)

**Reply by RHOOPH** [reply]
@patelharsh9797 

After i enabled luasnip extra, my useState snippet is working again, but it doesn't capitalize the setState part, 
[link to snippet](https://github.com/rafamadriz/friendly-snippets/blob/682157939e57bd6a2c86277dfd4d6fbfce63dbac/snippets/javascript/react-ts.json#L285)

Is this happening to you ?

**Reply by RHOOPH** [reply]
Solved it by installing luarocks and then installing jsregexp

https://github.com/kmarius/jsregexp

```
sudo luarocks --lua-version 5.1 install jsregexp
```

**Reply by patelharsh9797** [reply]
@RHOOPH  After enabling the luasnip extra everything is working fine for me. 
Make sure you have the right config for luasnip. I'm using the vs_code snippet with luasnip.

**Comment by francoisrob** [comment]
vtsls is pure garbage with js files, unreliable autocompletion in my experience. I also did the rollback to typescript-tools

**Reply by patelharsh9797** [reply]
How did you rollback?

**Reply by francoisrob** [reply]
> How did you rollback?

I apologize for the late reply.

I used the code this user linked [dpetka2001's comment](https://github.com/LazyVim/LazyVim/discussions/3603#discussioncomment-9749146)

Literally just copy pasted everything into a new lua file under plugins/ [here](https://github.com/francoisrob/dotfiles/blob/main/.config/nvim/lua/plugins/.ts.lua.bak)

I also had to copy some lazy extras manually to remove the typescript extra dependency. [seen here](https://github.com/francoisrob/dotfiles/blob/main/.config/nvim/lua/plugins/.angular.lua.bak)

**Comment by skriems** [comment]
I seem to have the same problems with vtsls. At my workplace I find myself in a pure js project and I cannot work with lazyvim, because I cannot see the code with all these errors and warnings.
Will definitely switch to tsserver. Maybe there should be a dedicated extra for each of them and the user can choose.

**Comment by stefano-regosa-deel** [comment]
Solved adding https://github.com/pmizio/typescript-tools.nvim and disabling the rest

```lua
return {
  "neovim/nvim-lspconfig",
  opts = {
    -- make sure mason installs the server
    servers = {
      tsserver = {
        enabled = false,
      },
      ts_ls = {
        enabled = false,
      },
      vtsls = {
        enabled = false,
      },
    },
  },
}
`
